python/mlir_structured/dialects/indexing.py

```diff
@@ -366,10 +370,268 @@ def __class_getitem__(
+  ellipses = (i for i, elt in enumerate(idx) if elt is Ellipsis)
+  ellipsis_index = next(ellipses, None)
+  if ellipsis_index is not None:
+    if next(ellipses, None) is not None:
+      raise IndexError(
+          f"Multiple ellipses (...) not supported: {list(map(type, idx))}.")
+    colons = (slice(None),) * (rank - len_without_none)
+    idx = idx[:ellipsis_index] + colons + idx[ellipsis_index + 1:]
+  elif len_without_none < rank:
+    colons = (slice(None),) * (rank - len_without_none)
+    idx = tuple(idx) + colons
+  return idx
+
+
+def _indices_to_indexer(idx: Sequence[Any], in_shape: Sequence[int]):
+  """Processes sequence of index objects and constructs _Indexer with
+  corresponding indexing tensor and collapse dims (i.e., scatter/gather dims).
+
+  Args:
+    idx: Sequence (list or tuple) of slices, ellipses, Scalar, or Tensors.
+    in_shape: The shape of the tensor being indexed into.
+
+  Returns:
+    _Indexer object.
+
+  """
+  idx = _canonicalize_tuple_index(idx, len(in_shape))
+
+  in_axis = 0  # Current axis in input.
+  out_axis = 0  # Current axis in output, before collapsing. See below.
+  collapsed_dims: Sequence[int] = []
+  indices: List[Tensor] = []
+  newaxis_dims: Sequence[int] = []
+
+  for idx_i, idx_e in enumerate(idx):
+    if _is_scalar(idx_e) and _has_index_type(idx_e):
+      # Handle basic int indexes.
+      idx_e = _expand_dims(idx_e, (0,))
+      indices.append(idx_e)
+      collapsed_dims.append(in_axis)
+      in_axis += 1
+    elif isinstance(idx_e, slice):
+      # Handle slice indices
+      out_axis += 1
+      in_axis += 1
+    else:
+      raise IndexError(
+          f"Indexing mode not yet supported. Open a feature request!\n{idx}")
+
+  collapsed_dims: Tuple[int, ...] = tuple(sorted(collapsed_dims))
+
+  if len(indices) == 1:
+    indices_tensor = indices[0]
+  else:
+    indices_tensor = concatenate(
+        indices,
+        0,
+    )
+
+  lit = indices_tensor.literal_value
+  # flatten all but last dim (i.e., idx/coord dim)
+  coords = lit.reshape(-1, lit.shape[-1])
+  unique_indices = len(np.unique(coords, axis=0)) == len(coords)
+
+  return _Indexer(
+      newaxis_dims=tuple(newaxis_dims),
+      collapsed_dims=collapsed_dims,
+      indices=indices_tensor,
+      unique_indices=unique_indices,
+  )
+
+
+def gather(
+    operand,
+    indices,
+    gather_dims,
+    *,
+    unique_indices: bool = False,
+):
+  return Tensor(
+      GatherOp(
+          source=operand,
+          indices=indices,
+          gather_dims=gather_dims,
+          unique=unique_indices,
+      ))
+
+
+def build_gather(
+    tens,
+    idx,
+    unique_indices=False,
+):
+  # Early bail for most trivial corner case (all full slices)
+  if all(i == slice(None) for i in idx):
+    return tens
+
+  indexer = _indices_to_indexer(idx, tens.shape)
+  out = tens
+
+  # We avoid generating a gather when indexer.gather_indices.size is empty.
+  if not _is_empty(indexer.indices):
```

Comment:
I think the logic would be simpler to test for `_is_empty` which `return tens` as another early exit and then handle the general case after that...

Reply:
lol true